internal/environmentscleanup/service.go

```diff
@@ -189,39 +192,30 @@ func (s *Service) getStaleRuntimesByShoots(labelSelector string) ([]runtime, []u
 		}
 
 		log.Infof("Shoot %q is older than %f hours with age: %f hours", shoot.GetName(), s.MaxShootAge.Hours(), shootAge.Hours())
-		staleRuntime, err := s.shootToRuntime(shoot)
-		if err != nil {
-			s.logger.Infof("found a shoot without kcp labels: %v", shoot.GetName())
-			shoots = append(shoots, shoot)
+
+		instances, _, _, _ := s.instanceStorage.List(dbmodel.InstanceFilter{Shoots: []string{shoot.GetName()}})
+		if len(instances) == 1 {
+			s.logger.Infof("Found an instance %q for shoot %q", instances[0].InstanceID, shoot.GetName())
+			staleRuntime := runtime{
+				ID:        instances[0].RuntimeID,
+				ShootName: shoot.GetName(),
+			}
+			runtimes = append(runtimes, staleRuntime)
 			continue
 		}
 
-		runtimes = append(runtimes, *staleRuntime)
-	}
-
-	return runtimes, shoots, nil
-}
-
-func (s *Service) shootToRuntime(st unstructured.Unstructured) (*runtime, error) {
-	shoot := gardener.Shoot{Unstructured: st}
-	runtimeID, ok := shoot.GetAnnotations()[shootAnnotationRuntimeId]
-	if !ok {
-		runtimeID, ok = shoot.GetAnnotations()[shootAnnotationInfrastructureManagerRuntimeId]
-		if !ok {
-			return nil, fmt.Errorf("shoot %q has no runtime-id annotation", shoot.GetName())
+		runtimeCR, err := s.getRuntimeCR(shoot.GetName())
```

Comment:
Typically there are less than 10 instances to remove, and most of them are provisioned using KEB (in this case this function is not called). I can't fetch by shoot name, because field selector is not supported.